Create new course general sidebar

onubrooks opened the pull request:
Fixes #6028 

### Changes proposed in this Pull Request

* Adds a 'General' panel to the new Sensei settings sidebar
* Adds a new item for each meta box that is displayed in the course settings
* Maintains the same meta-boxes when classic editor is used
* Adds a rest API route to update course teacher when post is updated

### Testing instructions

- Switch to the classic editor, 
- Edit a course and notice that these meta-boxes are displayed on the sidebar: teacher, course prerequisite, featured course, notifications etc
- Switch back to the block editor, a course.
- Click on Sensei Settings in the sidebar.
- Notice the 'General' panel opened. 
- Check that all the items in the meta-box from the classic editor are in this panel: teacher, course prerequisite, featured course, notifications.


### Known Bugs
Saving the course teacher works fine with the custom rest API. 
However, I noticed a strange bug where updating the post doesn't persist the new values set in the post meta fields using: 
`[ meta, setMeta ] = useEntityProp( 'postType', 'course', 'meta' );`, 
I may need a fresh pair of eyes to have a look.

### Screenshots

<img width="422" alt="Screenshot 2022-11-08 at 14 29 05" src="https://user-images.githubusercontent.com/26160845/200575514-5ef2d892-14d2-468f-a3aa-3730d0f7303c.png">

UPDATE: The issue with saving meta has been fixed.

fjorgemota committed:
Fix issue with enrolling students in the course view in a course without students

donnapep committed:
Add Course Overview block

donnapep committed:
Add Course Overview block to Course List block patterns

donnapep committed:
Add tests for block rendering

donnapep committed:
Fix comment

donnapep committed:
Fix Course List buttons extending outside container

donnapep committed:
Simplify CSS

fjorgemota committed:
Replace $$next-version$$ tags with 4.8.0

fjorgemota committed:
Update package.json version to 4.8.0

fjorgemota committed:
Update changelog to 4.8.0

fjorgemota committed:
Bump Plugin Header to 4.8.0

fjorgemota committed:
Update changelog on readme.txt

fjorgemota committed:
Support WP 6.1 and Update Stable Tag on readme.txt

fjorgemota committed:
Update POT file to 4.8.0

fjorgemota committed:
Fix PHP notice on Sensei Home Notices Provider

So, when neither level nor style is available, it doesn't emit a PHP notice.

fjorgemota committed:
If link on Sensei Home is internal, open on the same tab

Instead of opening new tabs for everything.

fjorgemota committed:
Fix broken test related to the Quick Links section

fjorgemota committed:
Merge pull request #6017 from Automattic/fix/php-notice-on-sensei-home-notices-provider

Fix PHP notice on Sensei Home Notices Provider

fjorgemota committed:
Don't open URLs in the task list in a new tab

fjorgemota committed:
Create isUrlExternal helper function on utils.js

fjorgemota committed:
Fix JSDoc types on TaskItem

fjorgemota committed:
Only add target="_blank" on task list when the link is external

renatho committed:
Fix plugin status after installed

renatho committed:
Merge pull request #6020 from Automattic/fix/status-after-installation

Fix status after installation

jom committed:
Add dismissible flag to admin notices

jom committed:
Add our padding to all Sensei Notices

jom committed:
Merge pull request #6022 from Automattic/add/dismissible-flag

Add support for dismissible flag to admin notices

jom committed:
Provide both plugin slug options for available updates

jom committed:
Just use the directory from the plugin file for the plugin slug

renatho committed:
Add internal animation to progress bar

fjorgemota committed:
Spacing fixes on the Sensei Pro Ad

renatho committed:
Merge pull request #6031 from Automattic/change/progress-bar-animation

Change progress bar animation

fjorgemota committed:
Adjust Sensei Pro Ad to refer to social quote instead

fjorgemota committed:
optimize SVG Sensei Circle Logo using svgo

fjorgemota committed:
Add rel="noreferrer" to links when the link is external

fjorgemota committed:
Remove the max-width from the author block on Sensei Pro Ad

fjorgemota committed:
Use min-height instead of height so the block can expand on mobile

fjorgemota committed:
Merge pull request #6019 from Automattic/change/open-internal-links-on-same-tab-sensei-home

fjorgemota committed:
Update changelog of the version 4.8.0

So we don't need to include the list containing all PRs related to this project

fjorgemota committed:
Update changelog on readme

jom committed:
Merge pull request #6029 from Automattic/fix/avail-updates-check

jom committed:
Only append name to module terms

jom committed:
Ensure we're looking at something that could be a user ID in module slug

fjorgemota committed:
Merge pull request #6032 from Automattic/change/sensei-pro-ad-adjustments

fjorgemota committed:
Do not reverse columns on mobile, showing blocks in the correct order

fjorgemota committed:
Expand buttons on mobile

fjorgemota committed:
Fix padding of the Sensei Pro Ad on mobile

jom committed:
Add back some packages from package-lock.json

jom committed:
Merge pull request #6040 from Automattic/fix/package-lock-json-480-release

Add back some packages from package-lock.json

onubrooks committed:
fix 6024: register new plugin sidebar for sensei

fjorgemota committed:
Merge pull request #6036 from Automattic/change/fix-sensei-pro-ad-on-mobile

Some fixes for the Sensei Pro Ad on mobile

fjorgemota committed:
Change Sensei documentation link to senseilms.com/docs

fjorgemota committed:
Update i18n files with latest changes to strings and file numbers

fjorgemota committed:
Fix occurence of wrong next-version tag

fjorgemota committed:
Remove old changelog entries from the readme.txt

onubrooks committed:
fix 6024: add sensei settings document settings panel, open plugin sidebar when clicked

fjorgemota committed:
Merge pull request #6011 from Automattic/release/4.8.0

onubrooks committed:
fix 6026: Move the current sensei course settings to the plugin sidebar

onubrooks committed:
refactor code

renatho committed:
Fix contact teacher block check

fjorgemota committed:
Add message for users without javascript enabled on Sensei Home

fjorgemota committed:
Delete unused admin view

fjorgemota committed:
Improve messaging when there's no JS available

donnapep committed:
Merge pull request #6010 from Automattic/fix/course-list-buttons

Fix Course List buttons extending outside container

LavaGolem committed:
Fix PHP notice in Course List Filter block

LavaGolem committed:
Merge pull request #6069 from Automattic/fix/fix_PHP_notice_in_Course_List_Filter_block

Fix PHP course list block notice on category archive view

aaronfc committed:
Merge pull request #5583 from Automattic/fix/enroll-students-course-view

Fix issue with enrolling students in the course view in a course with no students

onubrooks committed:
create new course general sidebar

jom committed:
Merge pull request #6034 from Automattic/fix/6033-wp61-cache-notice

Fix some issues with author names getting appended to module names

renatho committed:
Instantiate Contact Teacher block if Learning Mode is enabled

alexsanford committed:
Fix cursor on Reset Progress link

alexsanford committed:
Use defined constants

alexsanford committed:
Only start course progress if it hasn't been started already

renatho committed:
Avoid defining already defined constants

aaronfc committed:
Prevent multiple actions being enqueued at the same time.

- Previously first thing when running a job with Action Scheduler was to
  enqueue the next single_action.
- This behaviour made that next requests to the site could trigger a job
  concurrently.
- Some jobs like `Sensei_Enrolment_Course_Calculation_Job` did some
  process in batch relying on a "last_user_id" counter. These jobs would
  collide when running in parallel consuming resources by repeating the
  same user IDs multiple times.
- New approach enqueues a fallback action 60 seconds in the future first
  (so in case execution breaks we will re-try) and then right after
  completing the process clears enqueued actions and re-enqueues but
  this time immediately.
- We are enqueuing immediately to take profit of the AS capability to
  run multiple jobs per request.

onubrooks committed:
fix linter errors

renatho committed:
Merge pull request #6080 from Automattic/fix/defined-constants

Avoid defining already defined constants

yscik commented on assets/js/admin/course-general-sidebar.js:
```diff
@@ -0,0 +1,142 @@
+				label={ __( 'Feature this course.', 'sensei-lms' ) }
+				checked={ featured === 'featured' }
+				onChange={ ( checked ) =>
+					setMeta( { _course_featured: checked ? 'featured' : '' } )
```
I think the issue is that `setMeta` is not merging but replacing based on the [example here](https://github.com/WordPress/gutenberg/blob/trunk/docs/how-to-guides/metabox.md#step-2-add-meta-block). So these should be:
`setMeta( { ...meta, _course_featured: checked ? 'featured' : '' } )`

onubrooks replied:
Hey @yscik this didn't work either. Perhaps you could test it on your machine.
After some digging, here's a few things I found:

1. Only `_course_featured ` and `_course_prerequisite` are affected, all other meta fields save successfully.
2. I was thinking it's related to protected meta (those starting with '_') and that may be it, but `_course_theme` saves fine so I'm not sure.
3. `setMeta` seems to clear the values of `_course_featured ` and `_course_prerequisite` in the database. I set the correct values in the database, then view them in the editor and they look okay, however, after clicking on 'Update', the values are cleared again.

yscik commented on assets/js/admin/course-general-sidebar.js:
```diff
@@ -0,0 +1,142 @@
+	const prerequisite = meta._course_prerequisite;
+	const notification = meta.disable_notification;
+
+	const unsubscribe = subscribe( () => {
+		let editor = select( 'core/editor' );
```
There is a helper, [`editorLifecycle`](https://github.com/Automattic/sensei/blob/7c7a1c4a1244b5c5c11538dce4e38d44e28d18ae/assets/shared/helpers/editor-lifecycle.js#L17) that could work here. Also I think this should be in an effect and not done in every render. `editorLifecycle` will take care of running it after every save. Something like:

```
	useEffect( () => editorLifecycle( {
		onSave: () => {
			apiFetch( ... )
		},
	} ) );
```

yscik commented on assets/js/admin/course-general-sidebar.js:
```diff
@@ -0,0 +1,142 @@
+
+			<HorizontalRule />
+
+			<h3>Course Prerequisite</h3>
```
A few of the strings here are missing translation calls.

yscik commented on assets/js/admin/course-general-sidebar.js:
```diff
@@ -0,0 +1,142 @@
+			<h3>Course Prerequisite</h3>
+			{ ! courses.length ? (
+				<p>
+					{ ' ' }
```
Some extra whitespace around this.

yscik reviewed (commented):
👋 Only had a partial look at the meta stuff:

aaronfc committed:
Fix title encoding for course in task list.

aaronfc committed:
Fix border-radius for task list course preview.

onubrooks committed:
add some PHP unit tests

onubrooks committed:
Refactor: Use wp_add_inline_script, only call api if author was changed.

renatho committed:
Allow contact teacher block handle legacy lesson notices

onubrooks committed:
Refactor: useEffect so its not done in every render, add missing translation calls

aaronfc committed:
Use default parameter for immediate scheduling.

Co-authored-by: Renatho De Carli Rosa <renatho@gmail.com>

aaronfc committed:
Decode htmlentities for site title in Tasks List.

aaronfc committed:
Fix wrong variable name used.

- After last refactor a condition that was done later was being skipped
  because the variable name was changed. Reverting variable name to
  previous one.

aaronfc committed:
Some code cleanup for `Sensei_Quiz` class.

- Marking `$file` attribute as deprecated for later removal.
- Some minor phpdoc updates.
- Using strict comparison.
- Some comment typos.

aaronfc committed:
Fix typos in comments.

fjorgemota committed:
Make title of the Sensei Home page translatable again

alexsanford committed:
Fix Remove Progress link cursor

onubrooks committed:
some refactorings

alexsanford committed:
Fix link cursors for Students management for Lesson

fjorgemota committed:
Merge pull request #6059 from Automattic/change/sensei-home-no-js

onubrooks committed:
fix for linter

donnapep committed:
Merge pull request #5996 from Automattic/add/course-overview-block

Add Course Overview link to Course List block

onubrooks committed:
fix for linter

alexsanford committed:
Merge pull request #6079 from Automattic/fix/6064-start-date-reset

Fix start date reset

renatho committed:
Merge pull request #6058 from Automattic/fix/sensei-blocks

Fix contact teacher block check

renatho committed:
Merge pull request #6088 from Automattic/fix/random-question-set

Fix random question set displayed for answered quizzes.

aaronfc committed:
Disable sorting in User Profile view for Reports.

- Most of the columns are not sortable with the current implementation.
- Frontend was displaying them as sortable but sorting was not working
  correctly.
- Here we disable the sorting until we come up with a more robust way of
  implementing it.

aaronfc committed:
Disable sorting in internal Course views for Reports.

- Most of the columns are not sortable with the current implementation.
- Frontend was displaying them as sortable but sorting was not working
  correctly.
- Here we disable the sorting until we come up with a more robust way of
  implementing it.

aaronfc committed:
Disable sorting in Lesson internal view for Reports.

- Most of the columns are not sortable with the current implementation.
- Frontend was displaying them as sortable but sorting was not working
  correctly.
- Here we disable the sorting until we come up with a more robust way of
  implementing it.

jom committed:
Merge pull request #6083 from Automattic/fix/task-list

Some fixes for task list in Sensei Home

jom committed:
Merge pull request #6094 from Automattic/remove/broken-report-sorting

Disable broken sorting under Reports.

jom committed:
Bump version to 4.8.1

jom committed:
Add changelog for 4.8.1

jom committed:
Update POT file for 4.8.1

fjorgemota committed:
Use checkout@v3 instead of checkout@v2

To fix issue with NodeJS 12 being deprecated

fjorgemota committed:
Use cache@v3 instead of cache@v2

To fix issue with NodeJS 12 being deprecated.

fjorgemota committed:
Use upload-artifact@v3 instead of upload-artifact@v2

To avoid issue with NodeJS 12 being deprecated.

fjorgemota committed:
Avoid using set-output command because it is deprecated

jom committed:
Merge pull request #6096 from Automattic/release/4.8.1

Release 4.8.1

fjorgemota committed:
Always install latest npm 8 version

To keep avoiding NPM 9, but while still receiving the latest updates from NPM 8

aaronfc committed:
Consider missing update attributes.

- Some updates might come with missing attributes like `package`.
- Checked that `dirname` behaves correctly (empty result) with a `null`.

aaronfc committed:
Merge pull request #6089 from Automattic/fix/quiz-code-cleanup

Some code cleanup for `Sensei_Quiz` class.

aaronfc committed:
Merge pull request #6081 from Automattic/fix/action-scheduler-one-instance-at-time

Prevent multiple actions being enqueued at the same time.

aaronfc committed:
Merge pull request #6103 from Automattic/fix/notice-on-specific-updates

Fix warning when missing update attributes.

gikaragia commented:
This is strange, although it seems like the meta arguments are included in the request, they do not exist when the request is served in `class-wp-rest-server.php`. I'll further look into this next week.

![Screenshot 2022-11-11 at 18 07 10](https://user-images.githubusercontent.com/53191348/201382115-a845839d-3ff8-4911-85a1-615f131d12e6.png)

gikaragia commented:
Ok something else is happening. The REST request seems to be working after all. I managed to update the values momentarily but then they were cleared again. Maybe from a hook?

fjorgemota committed:
Merge pull request #6098 from Automattic/fix/github-actions-nodejs-version

yscik commented:
Looks like the problem is the metabox save handler:
https://github.com/Automattic/sensei/blob/d3149d91289738c49a84017a4f8e45db03753b68/includes/class-sensei-course.php#L62-L64

Those can be disabled now when using the block editor. ([WP_Screen::is_block_editor](https://developer.wordpress.org/reference/classes/wp_screen/is_block_editor/) could work for the check, unless it doesn't work when doing REST API requests?)

yscik commented:
There is also a small issue with the `Disable notifications on this course` option: when enabling the Custom Fields panel (in the block editor ⠇→ Preferences → Panels), that meta will show up (since its not private), and overwrite whatever is happening in the sidebar when saving.
Not sure it's worth to deal with this one.

onubrooks commented:
> Looks like the problem is the metabox save handler:
> 
> https://github.com/Automattic/sensei/blob/d3149d91289738c49a84017a4f8e45db03753b68/includes/class-sensei-course.php#L62-L64
> 
> Those can be disabled now when using the block editor. ([WP_Screen::is_block_editor](https://developer.wordpress.org/reference/classes/wp_screen/is_block_editor/) could work for the check, unless it doesn't work when doing REST API requests?)

Yup! That's it. Did a little test and it works without the hook.

onubrooks committed:
add check for block editor when saving metabox

onubrooks committed:
fix metabox hooks running after save

onubrooks committed:
Merge branch 'feature/course-settings-sidebar' into add/course-general-sidebar

onubrooks committed:
fix lint errors

onubrooks committed:
refactor logic

mikeyarce commented:
> > Looks like the problem is the metabox save handler:
> > https://github.com/Automattic/sensei/blob/d3149d91289738c49a84017a4f8e45db03753b68/includes/class-sensei-course.php#L62-L64
> > 
> > Those can be disabled now when using the block editor. ([WP_Screen::is_block_editor](https://developer.wordpress.org/reference/classes/wp_screen/is_block_editor/) could work for the check, unless it doesn't work when doing REST API requests?)
> 
> Yup! That's it. Did a little test and it works without the hook.

The only thing I would double-check is to make sure Bulk Actions and other actions still work from the non-gutenberg screens as I think some of them might also rely on these hooks.

onubrooks committed:
fix save meta bug

onubrooks commented:
Turns out `WP_Screen` class is not always available and when available the method `is_block_editor` throws an error (a method called statically and trying to access `$this`, its constructor is also private so can't be instantiated using `new` in this context).
Gutenberg calls `/wp-admin/post.php?post=8&action=edit&meta-box-loader=1&meta-box-loader-nonce=xxx` after every update which triggered the `save_post` hook. The real issue was with our saving logic here: 

`$new_meta_value = ( isset( $_POST[ $post_key ] ) ? sanitize_html_class( $_POST[ $post_key ] ) : '' );`

This guarantees that the meta will be saved regardless of whether or not it is present in `$_POST`. Fixed it in 9bb326ef0a8a3b0b9390196f17db6a89ed1c4dd6

onubrooks committed:
fix linter

mikeyarce commented:
This looks great, @onubrooks! 

I found two issues: 
- I still can't change the `Teacher` even wit your latest fix.  Something else I noticed is that I can't change the Teacher in the Classic Editor now either.
- In the Classic Editor, I can change enable "Feature this course" but if I disable it and update, it won't change the value.

Everything else in my tests worked great 🎉

onubrooks commented:
Thanks @mikeyarce 

> * I still can't change the `Teacher` even wit your latest fix.  Something else I noticed is that I can't change the Teacher in the Classic Editor now either.

Hmmm this one is tricky...
I found that if we add `author` to the course post type supports, it works fine: https://github.com/Automattic/sensei/blob/49e97ab5b6c506ec730582dc476d7806c8fefb98/includes/class-sensei-posttypes.php#L221 
The issue with this is that the author dropdown begins to show up in the quick and bulk edit menus. Can you play around and investigate if we can add author support and not have it show up in the quick edit forms? I did some digging but couldn't find a way.

UPDATE: added a possible fix in https://github.com/Automattic/sensei/pull/6077/commits/49dca9890f2bed73838f0d83b941a8d06d567f12. Please test again 🥲

onubrooks committed:
fix save teacher not working

mikeyarce commented:
Testing again - it works in the Classic Editor now 🎉 

But it still does not work in the Block Editor...  I'm trying to figure out why but I'm not sure right now.  What I do know is that it sometimes works.  If I change the teacher and save maybe 10 times, 2 of those times it will properly update, but the other times it won't.  That leads me to think that there's some kind of race condition happening?

mikeyarce commented:
Another clue...  I was watching the database while doing an update and I noticed that when I tried switching from Teacher 1 to Teacher 2... this happened:
1. Switches to Teacher 2
2. Switches back to Teacher 1

So there's some logic somewhere that's also updating the Teacher back to the original teacher after this new update happens.  But, sometimes it's after... which is why it sometimes changes.

mikeyarce commented:
More Clues!  If I block these requests from POSTing in the Network tab of my browser, it works every time:
```
http://senseifresh.local/wp-admin/post.php?post=131&action=edit&meta-box-loader=1&meta-box-loader-nonce=521c2f314b&_locale=user
http://senseifresh.local/wp-json/wp/v2/courses/131?_locale=user
```
I'm not totally sure how these work, but it looks like they are BOTH also trying to save the metadata here.

onubrooks commented:
> More Clues! If I block these requests from POSTing in the Network tab of my browser, it works every time:
> 
> ```
> http://senseifresh.local/wp-admin/post.php?post=131&action=edit&meta-box-loader=1&meta-box-loader-nonce=521c2f314b&_locale=user
> http://senseifresh.local/wp-json/wp/v2/courses/131?_locale=user
> ```
> 
> I'm not totally sure how these work, but it looks like they are BOTH also trying to save the metadata here.

Hey @mikeyarce 
The first URL saves metaboxes if there are any on the page. Anything that hooks into `save_post` gets called on that endpoint but validation and nonce verification kinda rules out `save_teacher_meta_box` in the class `Sensei_Teacher`.

The second URL is the default save endpoint when you click 'Update' in the editor. From what I gathered while doing some research, if there's no ['author' supports](https://github.com/Automattic/sensei/blob/49e97ab5b6c506ec730582dc476d7806c8fefb98/includes/class-sensei-posttypes.php#L221) for the post type (course in this case), the logged in user gets saved as the post author, this could be why our own API call (`/sensei-internal/v1/course-utils/update-teacher`) works but is later overwritten by another API call. https://github.com/Automattic/sensei/commit/49dca9890f2bed73838f0d83b941a8d06d567f12 fixed it for me though.

onubrooks committed:
save author in gutenberg update, phpcs fixes

mikeyarce commented:
@onubrooks - yup, it works now! 🎉 

- Classic Editor ✅ 
- Quick Edit ✅ 
- Block Editor ✅
